scipy/stats/mstats_basic.py

```diff
@@ -537,38 +548,76 @@ def kendalltau(x, y, use_ties=True, use_missing=False):
                 for i in range(len(ry)-1)], dtype=float)
     D = np.sum([((ry[i+1:] < ry[i])*(rx[i+1:] > rx[i])).filled(0).sum()
                 for i in range(len(ry)-1)], dtype=float)
+    xties = count_tied_groups(x)
+    yties = count_tied_groups(y)
     if use_ties:
-        xties = count_tied_groups(x)
-        yties = count_tied_groups(y)
         corr_x = np.sum([v*k*(k-1) for (k,v) in iteritems(xties)], dtype=float)
         corr_y = np.sum([v*k*(k-1) for (k,v) in iteritems(yties)], dtype=float)
         denom = ma.sqrt((n*(n-1)-corr_x)/2. * (n*(n-1)-corr_y)/2.)
     else:
         denom = n*(n-1)/2.
     tau = (C-D) / denom
 
-    var_s = n*(n-1)*(2*n+5)
-    if use_ties:
-        var_s -= sum(v*k*(k-1)*(2*k+5)*1. for (k,v) in iteritems(xties))
-        var_s -= sum(v*k*(k-1)*(2*k+5)*1. for (k,v) in iteritems(yties))
-        v1 = np.sum([v*k*(k-1) for (k, v) in iteritems(xties)], dtype=float) *\
-             np.sum([v*k*(k-1) for (k, v) in iteritems(yties)], dtype=float)
-        v1 /= 2.*n*(n-1)
-        if n > 2:
-            v2 = np.sum([v*k*(k-1)*(k-2) for (k,v) in iteritems(xties)],
-                        dtype=float) * \
-                 np.sum([v*k*(k-1)*(k-2) for (k,v) in iteritems(yties)],
-                        dtype=float)
-            v2 /= 9.*n*(n-1)*(n-2)
+    if method == 'exact' and (xties or yties):
+        raise ValueError("Ties found, exact method cannot be used.")
+    
+    if method == 'auto':
+        if (not xties and not yties) and (n <= 33 or min(C, n*(n-1)/2.0-C) <= 1):
+            method = 'exact'
         else:
-            v2 = 0
+            method = 'asymptotic'
+    
+    if not xties and not yties and method == 'exact':
+        # Exact p-value, see Maurice G. Kendall, "Rank Correlation Methods" (4th Edition), Charles Griffin & Co., 1970.
+        c = int(min(C, (n*(n-1))/2-C))
+        if n <= 0:
+            raise ValueError
+        elif c < 0 or 2*c > n*(n-1):
+            raise ValueError
+        elif n == 1:
+            prob = 1.0
+        elif n == 2:
+            prob = 1.0
+        elif c == 0:
+            prob = 2.0/np.math.factorial(n)
+        elif c == 1:
+            prob = 2.0/np.math.factorial(n-1)
+        else:
+            old = [0.0]*(c+1)
+            new = [0.0]*(c+1)
+            new[0] = 1.0
+            new[1] = 1.0
+            for j in range(3,n+1):
+                old = new[:]
+                for k in range(1,min(j,c+1)):
+                    new[k] += new[k-1]
+                for k in range(j,c+1):
+                    new[k] += new[k-1] - old[k-j]
+            prob = 2.0*sum(new)/np.math.factorial(n)
     else:
```

Comment:
We will now also end up in this else block for `method='somerandomtypo`. Making this `elif method == 'asymptotic'` and raising an error when `method` is an incorrect string would be good.

Reply:
Good point, fixed in the latest commit for both stats and mstats.